File: c/lc/path-with-minimum-effort.cc
// https://leetcode.com/contest/weekly-contest-212/problems/path-with-minimum-effort/
class Solution {
	struct Q {
		int cost;
		int x, y;
		int mx_cost;
		bool operator>(const Q &o) const {
			return cost > o.cost;		  
		}
	};
	struct P {
		int x,y;
	};
public:
	static int minimumEffortPath(const vector<vector<int>>& heights) {
		int h = heights.size();
		int w = heights[0].size();
		vector<vector<int>> cost(h, vector<int>(w, INT_MIN));
		priority_queue<Q, vector<Q>, greater<Q>> qq;
		qq.push(Q{0,0,0,0});
		while (!qq.empty()) {
			auto p = qq.top();
			cost[p.y][p.x] = max(p.cost, p.mx_cost);
			qq.pop();
			if (p.y == h-1 && p.x == w-1)
				break;
			for (P mv:(P[]){{-1,0}, {0,-1}, {1,0}, {0,1}}) {
				int x = p.x + mv.x;
				int y = p.y + mv.y;
				if (x>=0 && x<w && y>=0 && y<h && cost[y][x]==INT_MIN) {
					int cost = abs(heights[y][x]-heights[p.y][p.x]);
					qq.push(Q{cost, x, y, max(cost, p.mx_cost)});
				}
			}
		}
		return cost[h-1][w-1];
	}
};

File: kotlin/src/dk/leetcode/PathWithMinimumEffort.kt
package dk.leetcode
// https://leetcode.com/problems/path-with-minimum-effort/

import java.util.*

class PathWithMinimumEffort_Solution {
    private class D(val d: Int, val hfrom: Int, val t: Pair<Int, Int>)

    fun minimumEffortPath(mm: Array<IntArray>): Int {
        val h = mm.size
        val w = mm[0].size
        val dst = Pair(h - 1, w - 1)
        val seen = Array(h) { Array(w) { false } }
        val qq = PriorityQueue<D>() { d1, d2 -> if (d1.d < d2.d) -1 else +1 }
        qq.add(D(0, mm[0][0], Pair(0, 0)))
        while (qq.isNotEmpty()) {
            val nxt = qq.peek()
            qq.remove()
            if (nxt.t == dst)
                return nxt.d
            if (!seen[nxt.t.first][nxt.t.second]) {
                seen[nxt.t.first][nxt.t.second] = true
                for ((di, dj) in arrayOf(Pair(0, 1), Pair(-1, 0), Pair(0, -1), Pair(1, 0))) {
                    val i = nxt.t.first + di
                    val j = nxt.t.second + dj
                    if (i in 0 until h && j in 0 until w && !seen[i][j]) {
                        qq.add(D(Math.max(nxt.d, Math.abs(nxt.hfrom - mm[i][j])), mm[i][j], Pair(i, j)))
                    }
                }
            }
        }
        return -1
    }
}
